doc/fpmsyncd/hld_fpmsyncd-NTT.md

```diff
@@ -0,0 +1,383 @@
+Also, the `fpmsyncd` was modified to use the NextHop Group ID (`nexthop_group`) when programming the route to the `ROUTE_TABLE` if `RTA_NH_ID` was included in the `RTM_NEWROUTE` message from zebra via `FPM`.
+
+NHG ID and members are managed by `FRR`.
+`fpmsyncd` will use NHG ID provided in FPM message from `FRR(zebra)`.
+Thus, logic of either if updating NHG members or create NHG with new ID during topology change is managed by `FRR`.
+
+Use case example of this feature would be BGP PIC, and recursive routes handling.
+BGP PIC has started in design discussion in the SONiC Routing WG.
+Recursive routes support would be discussed after.
+See [09072023 Routing WG Meeting minutes](https://lists.sonicfoundation.dev/g/sonic-wg-routing/wiki/34786) for further information about BGP PIC discussion.
+
+### Requirements
+
+`Fpmsyncd extension` requires:
+- `fpmsyncd` to handle `RTM_NEWNEXTHOP` and `RTM_DELNEXTHOP` events from zebra via `dplane_fpm_nl`
+- `fpmsyncd` to SET/DEL routes to `APPL_DB: ROUTE_TABLE` using `nexthop_group`
+- `fpmsyncd` to SET/DEL NextHop Group entry to `APPL_DB: NEXTHOP_GROUP_TABLE`
+
+This feature must be disabled by default.
+- When this feature is disabled, behavior will be the same as before introducing this feature.
+  - i.e. `NEXTHOP_GROUP_TABLE` entry will not be created and `nexthop_group` will not be used in `ROUTE_TABLE` entry in `APPL_DB`.
+- See section [Configuration and management](#configuration-and-management) for details on how this feature is disabled/enabled.
+
+### Architecture Design 
+<!--
+
+This section covers the changes that are required in the SONiC architecture. In general, it is expected that the current architecture is not changed.
+This section should explain how the new feature/enhancement (module/sub-module) fits in the existing architecture. 
+
+If this feature is a SONiC Application Extension mention which changes (if any) needed in the Application Extension infrastructure to support new feature.
+-->
+
+This design modifies `fpmsyncd` to use the new `APPL_DB` tables.
+
+The current `fpmsyncd` handle just the `RTM_NEWROUTE` and `RTM_DELROUTE` writing all route information for each route prefix to `ROUTE_TABLE` on Redis DB (`redis-server`).
+When zebra process is initialized using the old fpm module, the `RTM_NEWROUTE` is sent with at least destination address, gateway, and interface id attributes.
+For multipath route, the `RTM_NEWROUTE` is sent with a list of gateway and interface id.
+
+This `Fpmsyncd extension` will modify `fpmsyncd` to handle  `RTM_NEWNEXTHOP` and `RTM_DELNEXTHOP` as below.
+
+<!-- omit in toc -->
+##### Figure: Fpmsyncd NHG High Level Architecture
+![fig: fpmsyncd nhg architecture](diagram/fpmsyncd-nhg-architecture.png)
+
+- FRR configuration
+  - (1) config zebra to use `dplane_fpm_nl` instead of `fpm` module (this is default since 202305 release)
+  - (2) set `fpm use-nexthop-groups` option (this is disabled by default and enabled via `CONFIG_DB`)
```

Comment:
does it assume the use of Unified Management Framework ?

Reply:
Is this [SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md](https://github.com/sonic-net/SONiC/blob/master/doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md) the Unified Management Framework you have asked?
If yes, we are using sonic-cfggen and `zebra.conf.j2` template to generate zebra config file based on CONFIG_DB entry.